diesel/src/connection/transaction_manager.rs

```diff
@@ -333,39 +333,37 @@ where
         let transaction_state = Self::get_transaction_state(conn)?;
 
         let rollback_sql = match transaction_state.in_transaction {
-            Some(ref mut in_transaction) => {
-                match in_transaction.transaction_depth.get() {
-                    1 => Cow::Borrowed("ROLLBACK"),
-                    depth_gt1 => {
-                        if in_transaction.top_level_transaction_requires_rollback {
-                            // There's no point in *actually* rolling back this one
-                            // because we won't be able to do anything until top-level
-                            // is rolled back.
-
-                            // To make it easier on the user (that they don't have to really look
-                            // at actual transaction depth and can just rely on the number of
-                            // times they have called begin/commit/rollback) we don't mark the
-                            // transaction manager as out of the savepoints as soon as we
-                            // realize there is that issue, but instead we still decrement here:
-                            in_transaction.transaction_depth = NonZeroU32::new(depth_gt1 - 1)
-                                .expect("Depth was checked to be > 1");
-                            return Ok(());
-                        } else {
-                            Cow::Owned(format!(
-                                "ROLLBACK TO SAVEPOINT diesel_savepoint_{}",
-                                depth_gt1 - 1
-                            ))
-                        }
-                    }
-                }
-            }
+            Some(ref mut in_transaction) => match in_transaction.transaction_depth.get() {
+                1 => Cow::Borrowed("ROLLBACK"),
+                depth_gt1 => Cow::Owned(format!(
+                    "ROLLBACK TO SAVEPOINT diesel_savepoint_{}",
+                    depth_gt1 - 1
+                )),
+            },
             None => return Err(Error::NotInTransaction),
         };
 
         match conn.batch_execute(&rollback_sql) {
             Ok(()) => {
-                Self::get_transaction_state(conn)?
-                    .change_transaction_depth(TransactionDepthChange::DecreaseDepth)?;
+                let tm_status = &mut Self::get_transaction_state(conn)?;
+                if let ValidTransactionManagerStatus {
+                    in_transaction:
+                        Some(InTransactionStatus {
+                            ref mut top_level_transaction_requires_rollback,
```

Comment:
We should probably rename this to top_level_transaction_may_require_rollback to explicit the definition that wouldn't match https://github.com/diesel-rs/diesel/issues/3470#issuecomment-1381598316 anymore